File: Leetcode/1218/1218.cpp
class Solution {
public:
    int longestSubsequence(vector<int>& arr, int difference) {
        unordered_map<int, vector<int>> unmap;
        int n = (int)arr.size();
        vector<int> dp(n, 1);
        int ans = 0;
        int val;
        // O(nlogn)
        for (int i = 0; i < n; ++i) {
            val = arr[i]-difference;
            if (unmap.find(val) != unmap.end() && i > unmap[val].front()) {
                auto iter = --lower_bound(unmap[val].begin(),unmap[val].end(),i);
                dp[i] = dp[*iter]+1;
            }
            unmap[arr[i]].emplace_back(i);
            ans = max(ans, dp[i]);
        }
        return ans;
    }
};

File: Leetcode/1218/1218.go
func longestSubsequence(arr []int, difference int) int {
	unmap := make(map[int][]int)
	dp := make([]int, len(arr))
	for i := 0; i < len(dp); i++ {
		dp[i] = 1
	}
	ans := 0
	for i := 0; i < len(dp); i++ {
		temp := arr[i] - difference
		val, ok := unmap[temp]
		if ok && i > val[0] {
			pos := sort.Search(len(val), func(j int) bool {
				return val[j] >= i
			})
			pos--
			dp[i] = dp[val[pos]] + 1
		}
		unmap[arr[i]] = append(unmap[arr[i]], i)
		if ans < dp[i] {
			ans = dp[i]
		}
	}
	return ans
}